### DISCUSSION TITLE
LateInitializationError

### DISCUSSION THREAD

**Discussion by kamyar75** [rrousselGit/riverpod#848] [General]
**Describe the bug**
LateInitializationError when sometimes use the first time run hookConsumerWidget.
just first run.
if I go to another page,and come back target page,soleved, but first time ...

**To Reproduce**

can't handle MapController

The following LateError was thrown while handling a gesture:
LateInitializationError: Field '_state' has not been initialized.

When the exception was thrown, this was the stack: 
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
packages/flutter_map/src/map/map.dart 16:23                                                                                get [_state]
packages/flutter_map/src/map/map.dart 64:31                                                                                get center
packages/mycity_sup/ui/app/app_setting_info_part.dart 822:47                                                               <fn>
packages/flutter/src/material/ink_well.dart 989:21           
**Expected behavior**
it need to run without error.

**Comment by TimWhiting** [comment]
If you look at the stack, it doesn't seem to be related to riverpod or hooks. It is the flutter_map package that is trying to access ._state without it being initialized. Check to make sure you are using that package properly, or provide a minimally reproducible example.

For example check to see if it happens with a stateful widget as well

**Comment by kamyar75** [comment]
with  stateful widget worked.

**Comment by TimWhiting** [comment]
Can you give a sample of what it looks like in StatefulWidget versus HookConsumerWidget, as well as what version of riverpod you are using?

**Comment by kamyar75** [comment]
I am using 1dev.11 version .

**Comment by rrousselGit** [comment]
As Tim mentioned, it's quite likely unrelated to hooks/riverpod.
Please share some code, so that we can help. Otherwise we cannot do anything

**Reply by kamyar75** [reply]
//****************************************************************************
//****************************************************************************
I have a problem with below code,when refresh in flutter web 
when just first time load
//****************************************************************************
//****************************************************************************

class AppSetMapCenterLoc extends HookConsumerWidget {
  AppSetMapCenterLoc({
    Key? key,
  }) : super(key: key);

  final MapController mapController = MapController();
  static double zoom = 14.0;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final appSettingsProvider = ref.watch(appSettingsChangeNotifierProvider);
    final appProvider = ref.read(appChangeNotifierProvider);
    //
    // final zoom = useState(14.0);
    // mapController = MapController();

    appSettingsProvider.mapController = mapController;

    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceAround,
          children: [
            Text(
              "مرکز نقشه شهر",
              style: Theme
                  .of(context)
                  .textTheme
                  .headline6,
            ),
            Expanded(
              child: Wrap(
                spacing: 10,
                runSpacing: 10,
                textDirection: TextDirection.ltr,
                crossAxisAlignment: WrapCrossAlignment.center,
                children: [
                  IconButton(
                      onPressed: () {
                        appSettingsProvider.refreshMap();
                      },
                      icon: const Icon(Ionicons.checkmark_circle_outline)),
                  SizedBox(
                    width: dpadding,
                  ),
                  SizedBox(
                    width: 100,
                    child: TextFormField(
                      controller: appSettingsProvider.tecLat,
                      style: Theme
                          .of(context)
                          .textTheme
                          .subtitle1,
                      keyboardType: TextInputType.number,
                      validator: (value) {
                        if (value!.isEmpty) {
                          return "طول جغرافیایی";
                        }

                        return null;
                      },
                      decoration: InputDecoration(
                          labelText: 'طول جغرافیایی',
                          border: AppTheme.inputBorder),
                    ),
                  ),
                  // if (Responsive.isMobile(context))
                  SizedBox(
                    height: dpadding,
                  ),
                  // if (Responsive.isMobile(context))
                  SizedBox(
                    width: 100,
                    child: TextFormField(
                      controller: appSettingsProvider.tecLong,
                      style: Theme
                          .of(context)
                          .textTheme
                          .subtitle1,
                      keyboardType: TextInputType.number,
                      validator: (value) {
                        if (value!.isEmpty) {
                          return "عرض جغرافیایی";
                        }

                        return null;
                      },
                      decoration: InputDecoration(
                          labelText: 'عرض جغرافیایی',
                          // errorText: ,
                          // filled: isExistPhone,

                          border: AppTheme.inputBorder),
                    ),
                  ),
                  SizedBox(
                    width: dpadding,
                  ),

                ],
              ),
            ),
          ],
        ),
        SizedBox(
          height: dpadding,
        ),
        SizedBox(
          height: 350,
          child: Stack(
            children: [
              ClipRRect(
                borderRadius: BorderRadius.circular(10),
                child: FlutterMap(
                  mapController: mapController,
                  options: MapOptions(
                    controller: mapController,
                    onTap: (tapPosition, point) {
                      appSettingsProvider.addMarkerToList(point);
                    },
                    center: appSettingsProvider.markerList.isNotEmpty
                        ? appSettingsProvider.markerList[0].point
                        : appProvider.centerMapLocation(),
                    minZoom: 9,
                    zoom: zoom,
                  ),
                  layers: [
                    appProvider.tileLayerOptions,
                    MarkerLayerOptions(markers: appSettingsProvider.markerList),
                  ],
                ),
              ),
              Positioned(
                  top: 0,
                  left: 0,
                  right: 0,
                  child: Padding(
                    padding: const EdgeInsets.all(12),
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: [
                        Row(
                          children: [
                            InkButton(
                              onPressed: () {
                                appSettingsProvider.getMyLocation();
                              },
                              child: Container(
                                  padding: EdgeInsets.all(5),
                                  decoration: BoxDecoration(
                                    color: Colors.black45,
                                    borderRadius:
                                    BorderRadius.circular(dpadding * 0.5),
                                  ),
                                  child: Row(
                                    children: [
                                      Icon(
                                        Ionicons.locate_outline,
                                        size: 24,
                                        color: Colors.white.withOpacity(0.7),
                                      ),
                                      const SizedBox(
                                        width: 5,
                                      ),
                                      Text(
                                        "مکان من",
                                        style: Theme
                                            .of(context)
                                            .textTheme
                                            .subtitle1!
                                            .copyWith(color: Colors.white),
                                      )
                                    ],
                                  )),
                            ),
                            const SizedBox(
                              width: 10,
                            ),
                          ],
                        ),
                      ],
                    ),
                  )),
              Positioned(
                  top: 0,
                  left: 5,
                  child: Padding(
                    padding: const EdgeInsets.all(12),
                    child: Row(
                      children: [
                        InkButton(
                          onPressed: () {
                            // print(zoom.value.toString());
                            zoom = zoom + 1.0;
                            mapController.move(mapController.center, zoom,
                                id: '11');
                            print(zoom.toString());
                          },
                          child: Container(
                            padding: const EdgeInsets.all(3),
                            decoration: BoxDecoration(
                              color: Colors.black45,
                              borderRadius:
                              BorderRadius.circular(dpadding * 0.5),
                            ),
                            child: Icon(
                              Ionicons.add,
                              size: 18,
                              color: Colors.white.withOpacity(0.7),
                            ),
                          ),
                        ),
                        const SizedBox(
                          width: 4,
                        ),
                        InkButton(
                          onPressed: () {
                            // print(zoom.value.toString());
                            zoom = zoom - 1;
                            mapController.move(mapController.center, zoom);
                            // print(zoom.value.toString());
                          },
                          child: Container(
                            padding: const EdgeInsets.all(3),
                            decoration: BoxDecoration(
                              color: Colors.black45,
                              borderRadius:
                              BorderRadius.circular(dpadding * 0.5),
                            ),
                            child: Icon(
                              Ionicons.remove,
                              size: 18,
                              color: Colors.white.withOpacity(0.7),
                            ),
                          ),
                        ),
                      ],
                    ),
                  )),
            ],
          ),
        ),
      ],
    );
  }
}

//****************************************************************************
//****************************************************************************
I moved my code to StatefulWidget and HookConsumer instede of HookConsumerWidget
my problem is solved. so, HookConsumerWidget has a problem.
//****************************************************************************
//****************************************************************************

class AppSetMapCenterLocStateful extends StatefulWidget {
  const AppSetMapCenterLocStateful({Key? key}) : super(key: key);

  @override
  _AppSetMapCenterLocStatefulState createState() =>
      _AppSetMapCenterLocStatefulState();
}

class _AppSetMapCenterLocStatefulState
    extends State<AppSetMapCenterLocStateful> {


  final MapController mapController = MapController();
  static double zoom = 14.0;


  @override
  Widget build(BuildContext context) {
    return HookConsumer(
      builder: (context, ref, child) {
        final appSettingsProvider = ref.watch(appSettingsChangeNotifierProvider);
        final appProvider = ref.read(appChangeNotifierProvider);
        //
        // final zoom = useState(14.0);
        // mapController = MapController();

        appSettingsProvider.mapController = mapController;

        return Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                Text(
                  "مرکز نقشه شهر",
                  style: Theme
                      .of(context)
                      .textTheme
                      .headline6,
                ),
                Expanded(
                  child: Wrap(
                    spacing: 10,
                    runSpacing: 10,
                    textDirection: TextDirection.ltr,
                    crossAxisAlignment: WrapCrossAlignment.center,
                    children: [
                      IconButton(
                          onPressed: () {
                            appSettingsProvider.refreshMap();
                          },
                          icon: const Icon(Ionicons.checkmark_circle_outline)),
                      SizedBox(
                        width: dpadding,
                      ),
                      SizedBox(
                        width: 100,
                        child: TextFormField(
                          controller: appSettingsProvider.tecLat,
                          style: Theme
                              .of(context)
                              .textTheme
                              .subtitle1,
                          keyboardType: TextInputType.number,
                          validator: (value) {
                            if (value!.isEmpty) {
                              return "طول جغرافیایی";
                            }

                            return null;
                          },
                          decoration: InputDecoration(
                              labelText: 'طول جغرافیایی',
                              border: AppTheme.inputBorder),
                        ),
                      ),
                      // if (Responsive.isMobile(context))
                      SizedBox(
                        height: dpadding,
                      ),
                      // if (Responsive.isMobile(context))
                      SizedBox(
                        width: 100,
                        child: TextFormField(
                          controller: appSettingsProvider.tecLong,
                          style: Theme
                              .of(context)
                              .textTheme
                              .subtitle1,
                          keyboardType: TextInputType.number,
                          validator: (value) {
                            if (value!.isEmpty) {
                              return "عرض جغرافیایی";
                            }

                            return null;
                          },
                          decoration: InputDecoration(
                              labelText: 'عرض جغرافیایی',
                              // errorText: ,
                              // filled: isExistPhone,

                              border: AppTheme.inputBorder),
                        ),
                      ),
                      SizedBox(
                        width: dpadding,
                      ),
                      // if (!Responsive.isMobile(context))
                      //   SizedBox(
                      //     width: 110,
                      //     child: TextFormField(
                      //       controller: appSettingsProvider.tecLong,
                      //       style: Theme.of(context).textTheme.subtitle1,
                      //       validator: (value) {
                      //         if (value!.isEmpty) {
                      //           return "عرض جغرافیایی";
                      //         }
                      //
                      //         return null;
                      //       },
                      //       decoration: InputDecoration(
                      //           labelText: 'عرض جغرافیایی',
                      //           // errorText: ,
                      //           // filled: isExistPhone,
                      //
                      //           border: OutlineInputBorder(
                      //               borderRadius: BorderRadius.circular(5),
                      //               borderSide: BorderSide(
                      //                 width: 1,
                      //                 color: Colors.black45,
                      //               ))),
                      //     ),
                      //   ),
                    ],
                  ),
                ),
              ],
            ),
            SizedBox(
              height: dpadding,
            ),
            SizedBox(
              height: 350,
              child: Stack(
                children: [
                  ClipRRect(
                    borderRadius: BorderRadius.circular(10),
                    child: FlutterMap(
                      mapController: mapController,
                      options: MapOptions(
                        controller: mapController,
                        onTap: (tapPosition, point) {
                          appSettingsProvider.addMarkerToList(point);
                        },
                        center: appSettingsProvider.markerList.isNotEmpty
                            ? appSettingsProvider.markerList[0].point
                            : appProvider.centerMapLocation(),
                        minZoom: 9,
                        zoom: zoom,
                      ),
                      layers: [
                        appProvider.tileLayerOptions,
                        MarkerLayerOptions(markers: appSettingsProvider.markerList),
                      ],
                    ),
                  ),
                  Positioned(
                      top: 0,
                      left: 0,
                      right: 0,
                      child: Padding(
                        padding: const EdgeInsets.all(12),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: [
                            Row(
                              children: [
                                InkButton(
                                  onPressed: () {
                                    appSettingsProvider.getMyLocation();
                                  },
                                  child: Container(
                                      padding: EdgeInsets.all(5),
                                      decoration: BoxDecoration(
                                        color: Colors.black45,
                                        borderRadius:
                                        BorderRadius.circular(dpadding * 0.5),
                                      ),
                                      child: Row(
                                        children: [
                                          Icon(
                                            Ionicons.locate_outline,
                                            size: 24,
                                            color: Colors.white.withOpacity(0.7),
                                          ),
                                          const SizedBox(
                                            width: 5,
                                          ),
                                          Text(
                                            "مکان من",
                                            style: Theme
                                                .of(context)
                                                .textTheme
                                                .subtitle1!
                                                .copyWith(color: Colors.white),
                                          )
                                        ],
                                      )),
                                ),
                                const SizedBox(
                                  width: 10,
                                ),
                              ],
                            ),
                          ],
                        ),
                      )),
                  Positioned(
                      top: 0,
                      left: 5,
                      child: Padding(
                        padding: const EdgeInsets.all(12),
                        child: Row(
                          children: [
                            InkButton(
                              onPressed: () {
                                // print(zoom.value.toString());
                                zoom = zoom + 1.0;
                                mapController.move(mapController.center, zoom,
                                    id: '11');
                                print(zoom.toString());
                              },
                              child: Container(
                                padding: const EdgeInsets.all(3),
                                decoration: BoxDecoration(
                                  color: Colors.black45,
                                  borderRadius:
                                  BorderRadius.circular(dpadding * 0.5),
                                ),
                                child: Icon(
                                  Ionicons.add,
                                  size: 18,
                                  color: Colors.white.withOpacity(0.7),
                                ),
                              ),
                            ),
                            const SizedBox(
                              width: 4,
                            ),
                            InkButton(
                              onPressed: () {
                                // print(zoom.value.toString());
                                zoom = zoom - 1;
                                mapController.move(mapController.center, zoom);
                                // print(zoom.value.toString());
                              },
                              child: Container(
                                padding: const EdgeInsets.all(3),
                                decoration: BoxDecoration(
                                  color: Colors.black45,
                                  borderRadius:
                                  BorderRadius.circular(dpadding * 0.5),
                                ),
                                child: Icon(
                                  Ionicons.remove,
                                  size: 18,
                                  color: Colors.white.withOpacity(0.7),
                                ),
                              ),
                            ),
                          ],
                        ),
                      )),
                ],
              ),
            ),
          ],
        );
      },
    );
  }
}

**Reply by rrousselGit** [reply]
Could you format your code properly? This isn't readable. Thanks :pray:

**Reply by kamyar75** [reply]
ok just some minute

**Reply by kamyar75** [reply]
//****************************************************************************
//****************************************************************************
I have a problem with below code,when refresh in flutter web
when just first time load
//****************************************************************************
//****************************************************************************

class AppSetMapCenterLoc extends HookConsumerWidget {
AppSetMapCenterLoc({
Key? key,
}) : super(key: key);

final MapController mapController = MapController();
static double zoom = 14.0;

@OverRide
Widget build(BuildContext context, WidgetRef ref) {
final appSettingsProvider = ref.watch(appSettingsChangeNotifierProvider);
final appProvider = ref.read(appChangeNotifierProvider);
//
// final zoom = useState(14.0);
// mapController = MapController();

appSettingsProvider.mapController = mapController;

return SizedBox(
      height: 350,
      child: Stack(
        children: [
          ClipRRect(
            borderRadius: BorderRadius.circular(10),
            child: FlutterMap(
              mapController: mapController,
              options: MapOptions(
                controller: mapController,
                onTap: (tapPosition, point) {
                  appSettingsProvider.addMarkerToList(point);
                },
                center: appSettingsProvider.markerList.isNotEmpty
                    ? appSettingsProvider.markerList[0].point
                    : appProvider.centerMapLocation(),
                minZoom: 9,
                zoom: zoom,
              ),
              layers: [
                appProvider.tileLayerOptions,
                MarkerLayerOptions(markers: appSettingsProvider.markerList),
              ],
            ),
          ),Positioned(
              top: 0,
              left: 5,
              child: Padding(
                padding: const EdgeInsets.all(12),
                child: Row(
                  children: [
                    InkButton(
                      onPressed: () {
                        // print(zoom.value.toString());
                        zoom = zoom + 1.0;
                        mapController.move(mapController.center, zoom,
                            id: '11');
                        print(zoom.toString());
                      },
                      child: Container(
                        padding: const EdgeInsets.all(3),
                        decoration: BoxDecoration(
                          color: Colors.black45,
                          borderRadius:
                          BorderRadius.circular(dpadding * 0.5),
                        ),
                        child: Icon(
                          Ionicons.add,
                          size: 18,
                          color: Colors.white.withOpacity(0.7),
                        ),
                      ),
                    ),
                    const SizedBox(
                      width: 4,
                    ),
                    InkButton(
                      onPressed: () {
                        // print(zoom.value.toString());
                        zoom = zoom - 1;
                        mapController.move(mapController.center, zoom);
                        // print(zoom.value.toString());
                      },
                      child: Container(
                        padding: const EdgeInsets.all(3),
                        decoration: BoxDecoration(
                          color: Colors.black45,
                          borderRadius:
                          BorderRadius.circular(dpadding * 0.5),
                        ),
                        child: Icon(
                          Ionicons.remove,
                          size: 18,
                          color: Colors.white.withOpacity(0.7),
                        ),
                      ),
                    ),
                  ],
                ),
              )),
        ],
      ),
    );
}
}

//****************************************************************************
//****************************************************************************
I moved my code to StatefulWidget and HookConsumer instede of HookConsumerWidget
my problem is solved. so, HookConsumerWidget has a problem.
//****************************************************************************
//****************************************************************************

class AppSetMapCenterLocStateful extends StatefulWidget {
const AppSetMapCenterLocStateful({Key? key}) : super(key: key);

@OverRide
_AppSetMapCenterLocStatefulState createState() =>
_AppSetMapCenterLocStatefulState();
}

class _AppSetMapCenterLocStatefulState
extends State {

final MapController mapController = MapController();
static double zoom = 14.0;

@OverRide
Widget build(BuildContext context) {
return HookConsumer(
builder: (context, ref, child) {
final appSettingsProvider = ref.watch(appSettingsChangeNotifierProvider);
final appProvider = ref.read(appChangeNotifierProvider);
//
// final zoom = useState(14.0);
// mapController = MapController();

    appSettingsProvider.mapController = mapController;

    return SizedBox(
      height: 350,
      child: Stack(
        children: [
          ClipRRect(
            borderRadius: BorderRadius.circular(10),
            child: FlutterMap(
              mapController: mapController,
              options: MapOptions(
                controller: mapController,
                onTap: (tapPosition, point) {
                  appSettingsProvider.addMarkerToList(point);
                },
                center: appSettingsProvider.markerList.isNotEmpty
                    ? appSettingsProvider.markerList[0].point
                    : appProvider.centerMapLocation(),
                minZoom: 9,
                zoom: zoom,
              ),
              layers: [
                appProvider.tileLayerOptions,
                MarkerLayerOptions(markers: appSettingsProvider.markerList),
              ],
            ),
          ),Positioned(
              top: 0,
              left: 5,
              child: Padding(
                padding: const EdgeInsets.all(12),
                child: Row(
                  children: [
                    InkButton(
                      onPressed: () {
                        // print(zoom.value.toString());
                        zoom = zoom + 1.0;
                        mapController.move(mapController.center, zoom,
                            id: '11');
                        print(zoom.toString());
                      },
                      child: Container(
                        padding: const EdgeInsets.all(3),
                        decoration: BoxDecoration(
                          color: Colors.black45,
                          borderRadius:
                          BorderRadius.circular(dpadding * 0.5),
                        ),
                        child: Icon(
                          Ionicons.add,
                          size: 18,
                          color: Colors.white.withOpacity(0.7),
                        ),
                      ),
                    ),
                    const SizedBox(
                      width: 4,
                    ),
                    InkButton(
                      onPressed: () {
                        // print(zoom.value.toString());
                        zoom = zoom - 1;
                        mapController.move(mapController.center, zoom);
                        // print(zoom.value.toString());
                      },
                      child: Container(
                        padding: const EdgeInsets.all(3),
                        decoration: BoxDecoration(
                          color: Colors.black45,
                          borderRadius:
                          BorderRadius.circular(dpadding * 0.5),
                        ),
                        child: Icon(
                          Ionicons.remove,
                          size: 18,
                          color: Colors.white.withOpacity(0.7),
                        ),
                      ),
                    ),
                  ],
                ),
              )),
        ],
      ),
    );
  },
);
}
}

**Reply by kamyar75** [reply]
two code, is same, just deffirence , use hookConsumerWidget  , use hookConsumer in stfulWidget

**Reply by kamyar75** [reply]
I think, hookConsumerWidget needs to initState same stfulWidget.
what is the lifesycle of HookConsumerWidget?

**Comment by kamyar75** [comment]
the problem in hookConsumerWidget when created init MapController.
but in stful this problem solved.

**Comment by chanjungkim** [comment]
@kamyar75 Have you solved this issue?
